Scaled tiles

tallytalwar opened the pull request:
Ok this is much better now.

- visible tile ids (scale and max zoom adjusted) are stored per datasource 
- Does not unnecessarily bloat TileID struct
- getParent and getChild TileID takes into account a datasource' tileScale and maxZoom respectively.

Addressed #1435 
- label pipeline modified a bit to include the set of datasources to help find parent and child tiles.

hjanetzek commented:
yup, looks much better :) 

Added some changes to keep the TileSource specific handling in TileManager instead of View::updateTiles

@matteblair, @tallytalwar your turn

tallytalwar commented:
@hjanetzek your changes look good to me.

tallytalwar commented on core/src/tile/tileManager.cpp:
```diff
@@ -154,8 +154,11 @@ void TileManager::updateTileSets(const View& _view) {
     // high to low zoom-levels.
-    std::sort(m_tiles.begin(), m_tiles.end(), [](auto& a, auto& b){
-            return a->getID() < b->getID(); });
+    std::sort(m_tiles.begin(), m_tiles.end(), [](auto& a, auto& b) {
```
👍

hjanetzek commented on core/src/tile/tileManager.cpp:
```diff
@@ -65,17 +65,22 @@ void TileManager::setTileSources(const std::vector<std::shared_ptr<TileSource>>&
+            } else {
+                LOGN("Reuse source from previous scene %s", source->name().c_str());
+                source = sIt->source;
+            }
```
We could also remove this TileSource reusing logic completely instead Imo.

tallytalwar replied:
Yeah, we were discussing this for https://github.com/tangrams/tangram-es/pull/1504.

tallytalwar replied:
Will address it in that PR.

matteblair commented on core/src/scene/sceneLoader.cpp:
```diff
@@ -971,6 +971,7 @@ void SceneLoader::loadSource(const std::shared_ptr<Platform>& platform, const st
     int32_t minDisplayZoom = -1;
     int32_t maxDisplayZoom = -1;
     int32_t maxZoom = 18;
+    int32_t tileScale = 0; // Equivalent of tileSize of 256px
```
More comment! 'A tileScale of 0 corresponds to a tile_size of 256, 1 corresponds to 512, 2 corresponds to 1024, etc.'

matteblair commented on core/include/tangram/tile/tileID.h:
```diff
@@ -45,28 +45,33 @@ struct TileID {
 
-    TileID getParent() const {
+    TileID scaled(int32_t _tileScale) const {
+        if (_tileScale <= 0) { return *this; }
```
Actually, maybe it makes more sense for these functions that require parameters from the tile source to be members of `TileSource` itself, like `TileID TileSource::getParent(TileID tile)`. Not sure that it helps substantially for this PR, but something to keep in mind perhaps.

tallytalwar replied:
Maybe, I won't consider it to be for this PR.

matteblair commented on core/src/view/view.h:
```diff
@@ -9,6 +9,7 @@
 #include "glm/vec3.hpp"
 #include <set>
 #include <memory>
+#include <unordered_map>
```
Unused also, I think. We can probably remove `<set>` now too.

matteblair commented on core/src/view/view.h:
```diff
@@ -133,7 +134,7 @@ class View {
     float getPitch() const { return m_pitch; }
 
-    /* Updates the view and projection matrices if properties have changed */
+    /* Updates the view and projection matrices if properties have changed  */
```
Sneaky space sneakin in

karimnaaji commented on core/src/scene/sceneLoader.cpp:
```diff
@@ -987,6 +988,15 @@ void SceneLoader::loadSource(const std::shared_ptr<Platform>& platform, const st
+    if (auto tileSizeNode = source["tile_size"]) {
+        const auto BaseTileSize = 256;
+        auto tileSize = tileSizeNode.as<int32_t>();
+        if (tileSize && !( tileSize & (tileSize - 1))) {
```
We do have a routine for checking power of two used for textures defined in `geom.h`, which would make this more explicit to read.

tallytalwar replied:
@karimnaaji will use the one you mentioned. @matteblair LOG statement below in the same context, one can easily tell what this statement is checking.

tallytalwar commented:
K @matteblair addressed most of of your suggestions. Few things I did not address:
1. TileSource::Options: makes sense, but does not directly relate to this PR and adds a lot of diff, so make sense to have a separate PR
2. getChildren. not sure if this is any beneficial to us, we have to loop over the children anyhow. And in unit test we operate with a single child! That said, I am neutral on this, but again does not directly relates to this PR and will add a lot of diff, so should be done separately.

hjanetzek commented on core/src/data/rasterSource.cpp:
```diff
@@ -69,8 +69,8 @@ class RasterTileTask : public BinaryTileTask {
-    : TileSource(_name, std::move(_sources), _minDisplayZoom, _maxDisplayZoom, _maxZoom),
+                           int32_t _, TextureOptions _options, bool _genMipmap)
+    : TileSource(_name, std::move(_sources), _minDisplayZoom, _maxDisplayZoom, _maxZoom, _),
       m_texOptions(_options),
```
Parameter name could be more descriptive :)

tallytalwar replied:
Ahh.. sed messed up. Will fix.

tallytalwar replied:
Updated.

hjanetzek commented on core/src/labels/labels.cpp:
```diff
@@ -214,23 +216,18 @@ void Labels::skipTransitions(const std::vector<std::unique_ptr<Style>>& _styles,
         TileID tileID = tile->getID();
         std::shared_ptr<Tile> proxy;
 
+        const auto& source = _scene->getTileSource(tile->sourceID());
```
Oh right, ClientDataSources are not added to the scene sources.

tallytalwar replied:
Yeah, this is fixed.

hjanetzek commented on core/src/scene/scene.cpp:
```diff
@@ -134,6 +134,15 @@ std::shared_ptr<Texture> Scene::getTexture(const std::string& textureName) const
     return texIt->second;
 }
 
+std::shared_ptr<TileSource> Scene::getTileSource(int32_t id) {
```
If something returns a pointer one should always check for null case :)

hjanetzek commented on core/src/labels/labels.cpp:
```diff
@@ -217,23 +221,19 @@ void Labels::skipTransitions(const std::vector<std::unique_ptr<Style>>& _styles,
 
+        const auto& source = _scene->getTileSource(tile->sourceID());
+        if (!source) { continue; }
+
```
no, ClientDataSources are not in Scene TileSources so this needs to be handled differently

tallytalwar commented:
No wait, @hjanetzek is correct. Labels skiptransition wont be checked for client datasources.

tallytalwar commented:
@hjanetzek can you check 382e479? I couldn't think of any other way to get client data sources accessed from within labels.

tallytalwar commented:
I took the scene file commit out from here, so this should be good to go.
